File: 209.minimum-size-subarray-sum.java
/*
 * @lc app=leetcode id=209 lang=java
 *
 * [209] Minimum Size Subarray Sum
 */

// @lc code=start
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        if (nums.length == 1) {
            return (nums[0] >= target) ? 1 : 0;
        }
        int left = 0;
        int subSum = 0;
        int winSize = Integer.MAX_VALUE;
        for (int right = 0; right != nums.length; ++right) {
            subSum += nums[right];
            while (subSum >= target) {
                winSize = Integer.min(winSize, right - left + 1);
                subSum -= nums[left];
                left += 1;
            }
        }
        return (winSize == Integer.MAX_VALUE) ? 0 : winSize;
    }
}
// @lc code=end

File: 209.minimum-size-subarray-sum.cpp
/*
 * @lc app=leetcode id=209 lang=cpp
 *
 * [209] Minimum Size Subarray Sum
 */

// @lc code=start

#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        auto l = nums.begin(), r = nums.begin() + 1;
        int subsum = nums.at(0);
        int minwinsz = INT_MAX;
        while (r <= nums.end()) {
            while (subsum >= target) {
                minwinsz = min(minwinsz, static_cast<int>(r - l));
                subsum -= *l;
                l += 1;
            }
            if (r == nums.end()) {
                break;
            }
            subsum += *r;
            r += 1;
        }
        return (minwinsz == INT_MAX) ? 0 : minwinsz;
    }
};
// @lc code=end
